Sources/Algorithms/Chain.swift

```diff
@@ -142,27 +143,27 @@ extension Chain: Collection where Base1: Collection, Base2: Collection {
   ) -> Index? {
     switch (i.position, limit.position) {
     case let (.first(i), .first(limit)):
-      let d = base1.distance(from: i, to: base1.endIndex)
-      if n < d {
+      if limit >= i {
+        // `limit` is relevant, so `base2` cannot be reached
         return base1.index(i, offsetBy: n, limitedBy: limit)
           .map(Index.init(first:))
+      } else if let j = base1.index(i, offsetBy: n, limitedBy: base1.endIndex) {
```

Comment:
I wish there was an operation that combined this calculation and finding `d` in the next branch, but no luck, right?

Reply:
I agree 100% to the extent that I added [the equivalent](https://github.com/rust-lang/rust/pull/76909) for Rust's `Iterator` recently, specifically to speed up `Chain` 🙂 This would make for a good optimization point if we ever get the chance.

The next best thing would be to write a regular function for this which somehow detects whether the collection supports random-access, is there any hope of that being possible to implement? I assume that info does exist at runtime but I haven't found a way to extract it.

Reply:
> I agree 100% to the extent that I added [the equivalent](https://github.com/rust-lang/rust/pull/76909) for Rust's `Iterator` recently, specifically to speed up `Chain` 🙂 This would make for a good optimization point if we ever get the chance.

Exactly. I think the only place we have anything like that is in the innards of capturing a sequence in an array, where we return the buffer along with in-progress iterator.
 
> The next best thing would be to write a regular function for this which somehow detects whether the collection supports random-access, is there any hope of that being possible to implement? I assume that info does exist at runtime but I haven't found a way to extract it.

There is a way, but since even non-random-access collections can provide a faster `index(_:offsetBy:)`, I don't think we'd want to go that route…